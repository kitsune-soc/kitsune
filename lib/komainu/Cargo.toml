[package]
name = "komainu"
authors.workspace = true
edition.workspace = true
version.workspace = true
license = "MIT OR Apache-2.0"

[[bench]]
name = "basic_auth"
harness = false

[[bench]]
name = "pkce"
harness = false

[dependencies]
base64-simd.workspace = true
bytes.workspace = true
compact_str.workspace = true
http.workspace = true
http-body.workspace = true
http-body-util.workspace = true
indexmap.workspace = true
itertools.workspace = true
memchr.workspace = true
serde.workspace = true
serde_urlencoded.workspace = true
sha2.workspace = true
simdutf8.workspace = true
sonic-rs.workspace = true
strum.workspace = true
subtle.workspace = true
thiserror.workspace = true
tracing.workspace = true
url.workspace = true

[dev-dependencies]
divan.workspace = true
futures-test.workspace = true
headers.workspace = true
insta.workspace = true
rstest.workspace = true
serde_test.workspace = true

[lints]
workspace = true
