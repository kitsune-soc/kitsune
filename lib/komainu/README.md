# komainu

![](https://www.japan-experience.com/sites/default/files/images/content_images/komainu20191.jpg)

Standards compliant generic OAuth2 server library for async Rust webservices.

Relevant standards:

- [RFC 6749](https://datatracker.ietf.org/doc/html/rfc6749)
- [RFC 7636](https://datatracker.ietf.org/doc/html/rfc7636)

## Goals

Make it easy for anyone to implement a full OAuth2 provider for their webservice.

## License

This project is licensed under either the MIT or Apache-2.0 license, at your choosing.
