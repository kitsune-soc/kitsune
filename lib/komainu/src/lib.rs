#[macro_use]
extern crate tracing;

use bytes::Bytes;
use std::collections::HashSet;
use std::{borrow::Cow, future::Future};
use strum::AsRefStr;

pub use self::error::{Error, Result};
pub use self::params::ParamStorage;

mod error;
mod params;

trait OptionExt<T> {
    fn or_missing_param(self) -> Result<T>;
}

impl<T> OptionExt<T> for Option<T> {
    #[inline]
    fn or_missing_param(self) -> Result<T> {
        self.ok_or(Error::MissingParam)
    }
}

// TODO: Refactor into `AuthorizerExtractor` and `Authorizer`
//
// `AuthorizerExtractor` contains the `ClientExtractor`, so we can load client info.
// `Authorizer` is the handle passed to the consumer to accept or deny the request.
// Unlike `oxide-auth`, we won't force the user to implement a trait here, the flow better integrates with a simple function.
//
// Because we use native async traits where needed, we can't box the traits (not that we want to), so at least the compiler can inline stuff well

pub struct Client<'a> {
    pub client_id: &'a str,
    pub client_secret: &'a str,
    pub scopes: Cow<'a, [Cow<'a, str>]>,
    pub redirect_uri: Cow<'a, str>,
}

pub trait ClientExtractor {
    fn extract(
        &self,
        client_id: &str,
        client_secret: Option<&str>,
    ) -> impl Future<Output = Result<Client<'_>>> + Send;
}

pub trait AuthIssuer {
    type UserId;

    fn issue_code(
        &self,
        user_id: Self::UserId,
        client_id: &str,
        scopes: &[&str],
    ) -> impl Future<Output = Result<String>> + Send;
}

#[derive(AsRefStr)]
#[strum(serialize_all = "snake_case")]
pub enum OAuthError {
    InvalidRequest,
    UnauthorizedClient,
    AccessDenied,
    UnsupportedResponseType,
    InvalidScope,
    ServerError,
    TemporarilyUnavailable,
}

#[inline]
fn get_from_either<'a>(
    key: &str,
    left: &'a ParamStorage<&str, &str>,
    right: &'a ParamStorage<&str, &str>,
) -> Option<&'a str> {
    left.get(key).or_else(|| right.get(key)).map(|item| &**item)
}

pub struct AuthorizerExtractor<AI, CE> {
    // pls do not use ai for this, even if the type alias implies it.
    // kthx bestie. bussi aufs bauchi.
    auth_issuer: AI,
    client_extractor: CE,
}

impl<AI, CE> AuthorizerExtractor<AI, CE>
where
    CE: ClientExtractor,
{
    pub fn new(auth_issuer: AI, client_extractor: CE) -> Self {
        Self {
            auth_issuer,
            client_extractor,
        }
    }

    pub async fn extract<'a>(&'a self, req: &'a http::Request<()>) -> Result<Authorizer<'a, AI>> {
        let query: ParamStorage<&str, &str> =
            serde_urlencoded::from_str(req.uri().query().or_missing_param()?)
                .map_err(Error::query)?;

        // TODO: Load client and verify the parameters (client ID, client secret, redirect URI, scopes, etc.) check out
        // Error out if that's not the case
        //
        // Check the grant_type, let the client access it _somehow_
        //
        // Give the user some kind of "state" parameter, preferably typed, so they can store the authenticated user, and their
        // consent answer.

        let client_id = query.get("client_id").or_missing_param()?;
        let response_type = query.get("response_type").or_missing_param()?;
        if *response_type != "code" {
            debug!(?client_id, "response_type not set to \"code\"");
            return Err(Error::Unauthorized);
        }

        let scope = query.get("scope").or_missing_param()?;
        let redirect_uri = query.get("redirect_uri").or_missing_param()?;
        let state = query.get("state").map(|state| &**state);

        let client = self.client_extractor.extract(client_id, None).await?;

        if client.redirect_uri != *redirect_uri {
            debug!(?client_id, "redirect uri doesn't match");
            return Err(Error::Unauthorized);
        }

        let request_scopes = scope.split_whitespace().collect::<HashSet<_>>();
        let client_scopes = client
            .scopes
            .iter()
            .map(|scope| &**scope)
            .collect::<HashSet<_>>();

        if !request_scopes.is_subset(&client_scopes) {
            debug!(?client_id, "scopes aren't a subset");
            return Err(Error::Unauthorized);
        }

        Ok(Authorizer {
            auth_issuer: &self.auth_issuer,
            client,
            query,
            state,
        })
    }
}

pub struct Authorizer<'a, AI> {
    auth_issuer: &'a AI,
    client: Client<'a>,
    query: ParamStorage<&'a str, &'a str>,
    state: Option<&'a str>,
}

impl<'a, AI> Authorizer<'a, AI>
where
    AI: AuthIssuer,
{
    pub fn client(&self) -> &Client<'a> {
        &self.client
    }

    pub fn query(&self) -> &ParamStorage<&'a str, &'a str> {
        &self.query
    }

    pub async fn accept(self, user_id: AI::UserId, scopes: &[&str]) -> http::Response<()> {
        // TODO: Call an issuer to issue an access token for a particular user
        // Construct the callback url
        // Construct a redirect HTTP response UwU

        let code = self
            .auth_issuer
            .issue_code(user_id, self.client.client_id, scopes)
            .await
            .unwrap();

        todo!();
    }

    pub async fn deny(self) -> http::Response<()> {
        todo!();
    }
}
