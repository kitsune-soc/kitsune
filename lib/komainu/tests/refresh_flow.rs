use self::fixtures::{Fixture, RNG_SEED, TokenValue, generate_secret};
use komainu::{
    ClientExtractor,
    flow::{self, SuccessTokenResponse, TokenType},
    scope::Scope,
};

mod fixtures;

#[futures_test::test]
async fn success_basic() {
    fastrand::seed(RNG_SEED);

    let fixtures = Fixture::generate();
    let client = fixtures
        .client_extractor
        .extract("client_1", None)
        .await
        .unwrap();

    let original_access_token = generate_secret();
    let refresh_token = generate_secret();

    fixtures.token_storage.insert(TokenValue {
        authorization: komainu::Authorization {
            code: generate_secret().into(),
            client: client.clone().into_owned(),
            pkce_payload: None,
            scopes: Scope::new(),
            user_id: "user_id".into(),
        },
        response: SuccessTokenResponse {
            access_token: original_access_token.clone().into(),
            token_type: TokenType::Bearer,
            refresh_token: refresh_token.clone().into(),
            expires_in: 1337,
        },
    });

    let body = sonic_rs::json!({
        "grant_type": "refresh_token",
        "refresh_token": refresh_token,

        "client_id": client.client_id,
        "client_secret": client.client_secret,
        "redirect_uri": client.redirect_uri,
    });
    let body = sonic_rs::to_string(&body).unwrap();

    let req = http::Request::builder()
        .header(http::header::CONTENT_TYPE, "application/json")
        .body(body)
        .unwrap();
    let req = komainu::Request::read_from(req).await.unwrap();

    let response = flow::refresh::perform(&req, &fixtures.client_extractor, &fixtures.refresh_flow)
        .await
        .unwrap();

    let success_response = response.unwrap_success();
    assert_ne!(success_response.access_token, original_access_token);

    insta::assert_json_snapshot!(success_response);
}

#[futures_test::test]
async fn invalid_grant_type() {
    fastrand::seed(RNG_SEED);

    let fixtures = Fixture::generate();
    let client = fixtures
        .client_extractor
        .extract("client_1", None)
        .await
        .unwrap();

    let body = sonic_rs::json!({
        "grant_type": "other grant type",
        "refresh_token": generate_secret(),

        "client_id": client.client_id,
        "client_secret": client.client_secret,
        "redirect_uri": client.redirect_uri,
    });
    let body = sonic_rs::to_string(&body).unwrap();

    let req = http::Request::builder()
        .header(http::header::CONTENT_TYPE, "application/json")
        .body(body)
        .unwrap();
    let req = komainu::Request::read_from(req).await.unwrap();

    let error = match flow::refresh::perform(
        &req,
        &fixtures.client_extractor,
        &fixtures.refresh_flow,
    )
    .await
    {
        Ok(..) => unreachable!(),
        Err(err) => err,
    };

    assert!(matches!(error, flow::Error::UnsupportedGrantType));
}
