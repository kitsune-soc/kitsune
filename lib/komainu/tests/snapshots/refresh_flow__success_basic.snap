---
source: lib/komainu/tests/refresh_flow.rs
expression: response
---
{
  "access_token": "kltugigbzriuurdb",
  "token_type": "Bearer",
  "refresh_token": "zisudtbgojtcgfsh",
  "expires_in": 1337
}
